crates/bevy_render/src/extract_component.rs

```diff
@@ -37,8 +37,26 @@ pub trait ExtractComponent: Component {
     type Query: WorldQuery + ReadOnlyWorldQuery;
     /// Filters the entities with additional constraints.
     type Filter: WorldQuery + ReadOnlyWorldQuery;
+
+    /// The output from extraction.
+    ///
+    /// Returning `None` based on the queried item can allow early optimization,
+    /// for example if there is an `enabled == false` field, or by only accepting
+    /// values within certain thresholds.
+    ///
+    /// The output may be different from the queried component.
+    /// This can be useful for example if only a subset of the fields are useful
+    /// in the render world.
+    ///
+    /// The output is a bundle instead of a component in order to allow use cases
+    /// such as tuples of components as output.
+    type Out: Bundle;
+
+    // TODO: https://github.com/rust-lang/rust/issues/29661
+    // type Out: Component = Self;
+
     /// Defines how the component is transferred into the "render world".
-    fn extract_component(item: QueryItem<'_, Self::Query>) -> Self;
+    fn extract_component(item: QueryItem<'_, Self::Query>) -> Option<Self::Out>;
```

Comment:
Do we want to add a `Clone` trait bound and add a default impl here?  `Some(item.clone())` seems to be the common pattern.

Reply:
Extracting uncloneable components generally seems impossible or at least poorly thought out?

Reply:
Hmm, I'm not too sure.

After these changes, the output type can be something entirely different than what is queried.
So a `Clone` bound on `Self` in that case is not strictly needed.